src/main/java/org/myrobotlab/framework/Service.java

```diff
@@ -3023,4 +2536,81 @@ public String getLocaleTag() {
     return locale.getTag();
   }
 
+  public boolean hasInterface(String interfaze) {
+    // probably a bad idea - but nice for lazy people
+    if (!interfaze.contains(".")) {
+      interfaze = String.format("org.myrobotlab.service.interfaces.%s", interfaze);
+    }
+    return interfaceSet.containsKey(interfaze);
+  }
+  
+  @Override
+  public boolean hasInterface(Class<?> interfaze) {    
+    return hasInterface(interfaze.getCanonicalName());
+  }
+  
+  public Set<String> getInterfaceNames() {
+    Set<String> fn = new TreeSet<>();    
+    Class<?>[] faces = getClass().getInterfaces();
+    for (Class<?> c : faces) {
+      fn.add(c.getCanonicalName());
+    }
+    return fn;
+  }
+
+  // a "helper" strongly typed Java function
+  @Override
+  public boolean isType(Class<?> clazz) {    
+    return isType(clazz.getCanonicalName());
+  }
+
+  /**
+   * This function does a type comparison of the service and a string
+   * passed in.  It's important that this does not use getClass() to resolve
+   * the type, instead to support polyglot proxies - it should be using a 
+   * string to compare types
+   */
+  @Override
+  public boolean isType(String clazz) {
+    // probably a bad idea - but nice for lazy people
+    if (!clazz.contains(".")) {
+      clazz = String.format("org.myrobotlab.service.%s", clazz);
+    }   
+    return serviceClass.equals(clazz);
+  }
+
+  @Override

```

Comment:
These method names don't make sense to me.  When are these called by the framework?  Does a service get an  "onCreated" called in addition to calling create ?  Or is this for peer services to notify a service...  very confusing.

Reply:
Any service has the opportunity to override and capture
onRegistered
onCreated
onStarted
onStopped
and
onReleased

registered and released have been events for a very long time.
The other callbacks were added to support create, start, and stopped.

One part of the complexity, is when a service starts and their are pre-existing services, should the starting service get existing service events?  I debated on this for a little, but the use case typically is fulfilling possibilities of other services for which it can attach.  If that is the most common use case, then its advantageous the new service see existing service events.  It makes a Service implementation of business logic less clean if it has to both wait for future events to handle in addition to query existing services manually.

Reply:
I feel like there should be another phase of the lifecycle between created and started..  or we need to consider that started will always load resources...  it seems like there is always a phase between the service being created.. where you are configuring it and loading resources.. and then finally you are starting the service ..